### PR TITLE
[WIP] Asset buyer 2.0 Dynamic swap quote consumer

### PR DESCRIPTION
**Pull request by dave4506**
## Description

Created a new consumer that will use the 'best' underlying consumer (Exchange, Forwarder, in the future Coordinator) to provide output or execute the swap quote. 

If the order is forwarder valid, the consumer will look at eth and weth balances as well to choose which consumer is best. 

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-buyer/src/quote_consumers/dynamic_swap_quote_consumer.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import { assetDataUtils } from '../utils/asset_data_utils';
+import { swapQuoteConsumerUtils } from '../utils/swap_quote_consumer_utils';
+
+import { ExchangeSwapQuoteConsumer } from './exchange_swap_quote_consumer';
+import { ForwarderSwapQuoteConsumer } from './forwarder_swap_quote_consumer';
+
+type SmartContractParams =
+    | ExchangeMarketBuySmartContractParams
+    | ExchangeMarketSellSmartContractParams
+    | ForwarderMarketBuySmartContractParams
```
Comment (fragosti) [inline packages/asset-buyer/src/quote_consumers/dynamic_swap_quote_consumer.ts]: Can we export these unions from their respective classes and then have the final union here.

File: packages/asset-buyer/src/quote_consumers/dynamic_swap_quote_consumer.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            const isEnoughEthAndWethBalance = _.map(takerEthAndWethBalance, (balance: BigNumber) =>
+                balance.isGreaterThanOrEqualTo(ethAmount),
+            );
+            if (isEnoughEthAndWethBalance[1]) {
+                // should be more gas efficient to use exchange consumer, so if possible use it.
+                return this._exchangeConsumer;
+            } else if (isEnoughEthAndWethBalance[0] && !isEnoughEthAndWethBalance[1]) {
+                return this._forwarderConsumer;
+            }
+            // Note: Should we default to a consumer if takerAddress has not enough Eth or Weth?
```
Comment (fragosti) [inline packages/asset-buyer/src/quote_consumers/dynamic_swap_quote_consumer.ts]: Maybe return the exchange consumer? Either way I agree we should return something and let that thing deal with the error. I would replace this comment with a comment that explains the reasoning.

File: packages/asset-buyer/src/quote_consumers/dynamic_swap_quote_consumer.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    public async executeSwapQuoteOrThrowAsync(
+        quote: SwapQuote,
+        opts: Partial<DynamicSwapQuoteExecutionOpts>,
+    ): Promise<string> {
+        assert.isValidSwapQuote('quote', quote);
+        const consumer = await this._getConsumerForSwapQuoteAsync(quote, opts);
+        return consumer.executeSwapQuoteOrThrowAsync(quote, opts);
+    }
+
+    private async _getConsumerForSwapQuoteAsync(
```
Comment (fragosti) [inline packages/asset-buyer/src/quote_consumers/dynamic_swap_quote_consumer.ts]: Is it tenable to write a test for this function?

Comment (dave4506) [reply packages/asset-buyer/src/quote_consumers/dynamic_swap_quote_consumer.ts]: making the function public; should make it testable

**Pull request merged**